src/main/java/org/spongepowered/common/command/WrapperCommandSource.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.command;
+
+import com.google.common.base.Optional;
+import net.minecraft.command.ICommandSender;
+import net.minecraft.util.Vec3;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.api.service.permission.MemorySubjectData;
+import org.spongepowered.api.service.permission.PermissionService;
+import org.spongepowered.api.service.permission.SubjectCollection;
+import org.spongepowered.api.service.permission.SubjectData;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.sink.MessageSink;
+import org.spongepowered.api.util.Tristate;
+import org.spongepowered.api.util.command.CommandMapping;
+import org.spongepowered.api.util.command.CommandSource;
+import org.spongepowered.api.util.command.source.EntitySource;
+import org.spongepowered.api.util.command.source.LocatedSource;
+import org.spongepowered.api.world.Location;
+import org.spongepowered.api.world.World;
+import org.spongepowered.common.Sponge;
+import org.spongepowered.common.interfaces.IMixinCommandSender;
+import org.spongepowered.common.service.permission.SpongePermissionService;
+import org.spongepowered.common.service.permission.base.SpongeSubject;
+import org.spongepowered.common.text.SpongeTexts;
+import org.spongepowered.common.util.VecHelper;
+
+public class WrapperCommandSource extends SpongeSubject implements CommandSource {
+
+    final ICommandSender sender;
+    private final MemorySubjectData data;
+
+    private WrapperCommandSource(ICommandSender sender) {
+        this.sender = sender;
+        this.data = new MemorySubjectData(Sponge.getGame().getServiceManager().provide(PermissionService.class).get());
```

Comment:
I think it would be preferable to delegate to permissions plugins for this if possible (but have to figure out how to do identifiers, and an appropriate subject type (maybe `minecraft`?)).